lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -151,14 +186,74 @@ class LatticeValue {
   bool operator!=(const LatticeValue &other) const {
     return valueAndTag != other.valueAndTag;
   }
+  friend raw_ostream &operator<<(raw_ostream &os, const LatticeValue &dt);
 
 private:
   /// The attribute value if this is a constant and the tag for the element
   /// kind.  The attribute is always an IntegerAttr.
   llvm::PointerIntPair<Attribute, 2, Kind> valueAndTag;
 };
+
+raw_ostream &operator<<(raw_ostream &os, const LatticeValue &lattice) {
+  os << "<unknown: " << lattice.isUnknown()
+     << ", invalid: " << lattice.isInvalidValue()
+     << ", constant: " << lattice.isConstant()
+     << ", over: " << lattice.isOverdefined() << ">";
+  return os;
+}
+
+// This class defines the data structure associated with lattice values.
+// This means basically a pair of value and integer. The integer
+// represents the relative offset to the value on leaf level of type.
+class ValueAndLeafIndex {
+public:
+  ValueAndLeafIndex(Value value, unsigned leafId, bool isKnownRoot = false)
+      : valueAndFlag(value, isKnownRoot), leafId(leafId) {}
```

Comment:
`isKnownRoot` encoding is used to avoid recomputation of the root value. But on the second thought, I think it is sufficient to check whether `isa<SubfieldOp, SubindexOp>` is true.

> I think this is equivalent to leafId == 0

`leafId==0` doesn't always indicate the value is root. For example,  
```
%a = reg : firrtl.vector<vector<uint, 3>, 3>
%a_0 = firrtl.subindex %a[0]
%a_0_0 = firrtl.subindex %a_0[0]
```
LeafIds of `%a, %a_0, %a_0_0` are all 0 because they point at first element. However, only `%a` must be root.